examples/EF_QAE/main.py

```diff
@@ -0,0 +1,246 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+import numpy as np
+from qibo.models import Circuit
+from qibo import hamiltonians, gates, models, matrices
+from qibo.hamiltonians import Hamiltonian
+from scipy.optimize import minimize
+from sklearn.datasets import load_digits
+import sys
+import argparse
+
+
+def main(layers, autoencoder, example):
+    
+    def encoder_hamiltonian_simple(nqubits, ncompress):
+        """Creates the encoding Hamiltonian.
+        Args:
+            nqubits (int): total number of qubits.
+            ncompress (int): number of discarded/trash qubits.
+
+        Returns:
+            Encoding Hamiltonian.
+        """
+        m0 = hamiltonians.Z(ncompress, numpy=True).matrix
+        m1 = np.eye(2 ** (nqubits - ncompress), dtype=m0.dtype)
+        ham = hamiltonians.Hamiltonian(nqubits, np.kron(m1, m0))
+        return 0.5 * (ham + ncompress)
+    
+    def ansatz_QAE(theta):
+        """Creates the variational quantum circuit for QAE.
+        Args:
+            theta (array or list): values of the parameters.
+
+        Returns:
+            Quantum circuit.
+        """
+        circuit = models.Circuit(nqubits)
+        index = 0
+        for l in range(layers):
+            for q in range(nqubits):
+                circuit.add(gates.RY(q, theta[index]))
+                index+=1
+            circuit.add(gates.CZ(5, 4)); circuit.add(gates.CZ(5, 3))
+            circuit.add(gates.CZ(5, 1)); circuit.add(gates.CZ(4, 2)); circuit.add(gates.CZ(4, 0))
+            for q in range(nqubits):
+                circuit.add(gates.RY(q, theta[index]))
+                index+=1
+            circuit.add(gates.CZ(5, 4)); circuit.add(gates.CZ(5, 2))
+            circuit.add(gates.CZ(4, 3)); circuit.add(gates.CZ(5, 0)); circuit.add(gates.CZ(4, 1))
+        for q in range(nqubits-compress, nqubits, 1):
+            circuit.add(gates.RY(q, theta[index]))
+            index+=1
+        return circuit
```

Comment:
As @stavros11 was mentioning, I also believe is much faster to allocate the ansatz circuits once and then use the `set_parameter` method during the optimization.

Reply:
How will this work on the other ansatz, that has two parameters on each Ry? I know how to do it in the one that you mentioned, but not in the other case.

Reply:
In principle you can write a function that takes `theta` from the optimizer, and rotate to the circuit basis, then you send this object with set_parameters.

Reply:
You can do something like this:
```python
def rotate(theta, x):
    new_theta = []
    for l in range(layers):
        for q in range(nqubits):
           new_theta.append(theta[i]*x + theta[index+1])
           index+=2
    ...
   return new_theta
```
and then the cost function should do
```python
newparams = rotate(params)
circuit.set_parameters(newparams)
final_state = circuit()
```